File: LeetCode/DP/Python/Leetcode467-unique-substrings-in-wraparound-string.py
class Solution:
    def findSubstringInWraproundString(self, p: str) -> int:
        mp = defaultdict(int)
        i = 0
        while i < len(p):
            j = i + 1
            while j < len(p) and ((ord(p[j]) - ord(p[j - 1]) == 1) or (p[j] == 'a' and p[j - 1] == 'z')): 
                print(j)
                j += 1
            while i < j: 
                mp[p[i]] = max(mp[p[i]],j - i)
                i += 1
        res = 0
        for k,v in mp.items():
            res += v
        return res

File: LeetCode/DP/Golang/Leetcode467-unique-substrings-in-wraparound-string.go
package sbgo

func findSubstringInWraproundString(p string) int {
	mp := map[byte]int{}
	for i := 0; i < len(p); {
		j := i + 1
		for j < len(p) && (p[j]-p[j-1] == 1 || (p[j] == 'a' && p[j-1] == 'z')) {
			j++
		}
		for i < j {
			mp[p[i]] = max(mp[p[i]], j-i)
			i++
		}
	}
	res := 0
	for _, v := range mp {
		res += v
	}
	return res
}

func max(i, j int) int {
	if i < j {
		return j
	}
	return i
}
